client/python/datajunction/nodes.py

```diff
@@ -0,0 +1,331 @@
+    deactivated_at: Optional[int]
+
+    @abc.abstractmethod
+    def _update(self) -> "Node":
+        """
+        Update the node for fields that have changed.
+        """
+
+    def save(self, mode: Optional[models.NodeMode] = models.NodeMode.PUBLISHED) -> dict:
+        """
+        Sets the node's mode to PUBLISHED and pushes it to the server.
+        """
+        existing_node = self.dj_client._get_node(node_name=self.name)
+        if "name" in existing_node:
+            # update
+            response = self._update()
+            if not response.ok:  # pragma: no cover
+                raise DJClientException(
+                    f"Error updating node `{self.name}`: {response.text}",
+                )
+            self.refresh()
+        else:
+            # create
+            response = self.dj_client._create_node(node=self, mode=mode)
+            if not response.ok:  # pragma: no cover
+                raise DJClientException(
+                    f"Error creating new node `{self.name}`: {response.text}",
+                )
+        return response.json()
+
+    def refresh(self):
+        """
+        Refreshes a node with its latest version from the database.
+        """
+        refreshed_node = self.dj_client._get_node(self.name)
+        for key, value in refreshed_node.items():
+            if hasattr(self, key):
+                setattr(self, key, value)
+        return self
+
+    def link_dimension(
+        self,
+        column: str,
+        dimension: str,
+        dimension_column: Optional[str],
+    ):
+        """
+        Links the dimension to this node via the node's `column` and the dimension's
+        `dimension_column`. If no `dimension_column` is provided, the dimension's
+        primary key will be used automatically.
+        """
+        link_response = self.dj_client._link_dimension_to_node(
+            self.name,
+            column,
+            dimension,
+            dimension_column,
+        )
+        self.refresh()
+        return link_response
+
+    def unlink_dimension(
+        self,
+        column: str,
+        dimension: str,
+        dimension_column: Optional[str],
+    ):
+        """
+        Removes the dimension link on the node's `column` to the dimension.
+        """
+        link_response = self.dj_client._unlink_dimension_from_node(  # pylint: disable=protected-access
+            self.name,
+            column,
+            dimension,
+            dimension_column,
+        )
+        self.refresh()
+        return link_response
+
+    def add_materialization(self, config: models.MaterializationConfig):
+        """
+        Adds a materialization for the node. This will not work for source nodes
+        as they don't need to be materialized.
+        """
+        upsert_response = self.dj_client._upsert_materialization(
+            self.name,
+            config,
+        )
+        self.refresh()
+        return upsert_response
+
+    def deactivate(self) -> str:
```

Comment:
To your question in the PR description, deactivate and activate is meant to act as a soft delete. This should appear to the user at the surface as a normal delete so maybe on the client side of things we scrap the deactivate/activate terminology for delete/restore.

Reply:
Thanks, I'll do these changes in my next PR.